# Pythonにも条件付きfor文が欲しい

これね
```cpp
for (int i = 2; i*i <= N; i++) {
    // 処理
}
```

In [10]:
N = 10000000

In [11]:
%%timeit

cnt = 0
for i in range(N):
    cnt += i

322 ms ± 13.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [12]:
%%timeit

cnt = 0
i = 0
while i <= N:
    cnt += i
    i += 1

565 ms ± 16.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [13]:
def range_while(start, condition):
    while condition(start):
        yield start
        start += 1

In [14]:
%%timeit

cnt = 0
for i in range_while(0, lambda i: i <= N):
    cnt += i

1.05 s ± 37.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [15]:
from itertools import count, takewhile

In [16]:
%%timeit

cnt = 0
for i in takewhile(lambda x: x<N, count()):
    cnt += i

757 ms ± 17.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


## 処理速度比較

| loop-type | 10^5 | 10^7 |
| :-------: | :--: | :--: |
| for loop | 3.12 ms ± 49.6 µs | 322 ms ± 13.5 ms |
| while loop | 5.67 ms ± 143 µs | 565 ms ± 16.9 ms |
| range_while | 10.4 ms ± 228 µs | 1.05 s ± 37.3 ms |
| itertools.takewhile | 7.38 ms ± 150 µs | 757 ms ± 17.6 ms |

## 結論
基本は`while`、どうしてもってときは`takewhile`

# 累積和は`itertools.accumulate`を採用すべき??

In [2]:
from itertools import accumulate

In [3]:
from random import randint

N = 100000
A = [randint(0, N) for _ in range(N)]

In [5]:
%%timeit

S = [0] * (N+1)
for i in range(N):
    S[i+1] = S[i] + A[i]

8.98 ms ± 59.5 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [8]:
%%timeit

S = list(accumulate(A))

1.98 ms ± 65.4 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [9]:
%%timeit

S = [0] + list(accumulate(A))

2.14 ms ± 17.6 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


## 結論
絶対に`accumulate`を採用しよう

# 